10-suspense-for-data-fetching/README.md

```diff
@@ -0,0 +1,148 @@
+- The APIs can change in the future.
+- Only Facebook it is using in production with Relay.
+- One of the things that I love about React in general, is that they're always applying the things that they're presenting to us internally at Facebook first, so they get some of the quirks worked out.
+- We're playing around with things, we're making little abstractions from things. Some of these things might be a bad idea, some of them might be good ideas. This is the whole point of this workshop, is to experiment with this, get familiar with the ideas and the concepts and the things that it enables, and then let you run off and play around with it, and give feedback back to the React team.
+
+## Approaches
+
+### Fetch-on-Render (not using Suspense)
+
+Start rendering components. Each of these components may trigger data fetching in their effects and life-cycle methods. This approach often leads to “waterfalls”.
+
+### Fetch-Then-Render (not using Suspense)
+
+Start fetching all the data for the next screen as early as possible. When the data is ready, render the new screen. We can’t do anything until the data arrives.
+
+1. Start fetching
+2. Finish fetching
+3. Start rendering
+
+### Render-as-You-Fetch (using Suspense)
+
+Start fetching all the required data for the next screen as early as possible, and start rendering the new screen immediately — before we get a network response. As data streams in, React retries rendering components that still need data until they’re all ready.
+
+  1. Start fetching
+  2. Start rendering
+  3. Finish fetching
+
+## Setup
+
+Use concurrent mode to enable experimental features.
+
+```
+ReactDOM.createRoot(document.getElementById('root')).render(<App />);
+```
+
+## Usage
+
+```
+// This is not a Promise. It's a special object (resource) from our Suspense integration.
+const peopleResource = wrapPromise(fetchPeople);
+
+const People = () => {
+  // Try to read people list, although it might not have loaded yet
+  const people = peopleResource.read();
+
+  return (
+    <ul>
+      {people.map((person, index) => (
+        <li key={index}>{person.name}</li>
+      ))}
+    </ul>
+  );
+};
+
+const Page = () => {
+  return (
+    <React.Suspense fallback={<div>Loading people...</div>}>
+      <People />
+    </React.Suspense>
+  );
+};
+```
+
+Here’s what happens when we render `<Page>` on the screen:
+
+- We’ve already kicked off the requests in `peopleResource`. It gave us a special “resource” instead of a Promise. In a realistic example, it would be provided by our data library’s Suspense integration, like Relay.
+
+- React tries to render `<Page>`. It returns `<People>` as children.
+
+- React tries to render `<People>`. It calls `peopleResource.read()`. None of the data is fetched yet, so this component “suspends”. React skips over it, and tries rendering other components in the tree.
```

Comment:
should we say how the suspension is implemented? `read()` will throw a Promise, so no further code in `<People>` is executed..

when React catches the exception, it will find out it is a Promise (instead of Error), so it will use nearest `fallback` and wait until it resolves to trigger a re-render (instead of triggering `componentDidCatch`)

Reply:
Yeah, I will talk about it. But I can add it here also.